documentation/specs/MultithreadedMSBuild/thread-safe-tasks.md

```diff
@@ -0,0 +1,267 @@
+
+Task authors who need to support older MSBuild versions have three choices:
+1. **Maintain separate implementations** - Create both thread-safe and legacy versions
+2. **Use compatibility bridge approaches** - Rely on MSBuild's ability to run legacy tasks in multithreaded mode without ExecutionContext access
+3. **Accept reduced performance** - Tasks will execute slower than their thread-safe version because they must run in a separate TaskHost process.
+
+For legacy tasks to run safely in multithreaded builds, they must:
+- Not modify environment variables or current working directory
+- Not depend on relative path resolution from current working directory
+- Not use static fields or other shared state without proper synchronization
+- Not rely on specific environment variable values being set
+
+We can implement any combinations of the options below. At least one of them should be the compatibility bridge option: compatible with older MSBuild versions. 
+
+### Option 1: Interface-Based Thread-Safe Declaration
+
+Task authors indicate thread-safety capabilities by implementing the `IThreadSafeTask` interface.
+
+**Pros:**
+- Clear, explicit opt-in mechanism for thread safety
+- Follows established MSBuild patterns (e.g., `ICancelableTask`)
+- Provides access to ExecutionContext APIs
+
+**Cons:**
+- Not compatible with older MSBuild versions. The task will not be able to load, so authors will need to create separate task classes, and even separate assemblies if a wide range of MSBuild versions needs to be supported.
+
+Task authors who want to support older MSBuild versions need to:
+- Maintain both thread-safe and legacy implementations
+- Use conditional task declarations based on MSBuild version to select which assembly to load the task from. This is necessary because task declarations don't support aliases - the task name serves as both the identifier in project files and the class name to locate in the assembly.
+- Handle version detection and selection logic
+
+**Note:** Consider backporting `IThreadSafeTask` to MSBuild 17.14 for graceful failure when the interface is used.
+
+**Option A:** Property-based approach
+```csharp
+public interface IThreadSafeTask : ITask
+{
+    ITaskExecutionContext ExecutionContext { get; set; }
+}
+```
+
+**Option B:** Parameter-based approach
+```csharp
+public interface IThreadSafeTask : ITask
+{
+    bool Execute(ITaskExecutionContext context);
+}
+```
+
+**Pros of Option A:**
+- Better feature decoupling for future extensibility. For example, if we add async support, we need only one method (`ExecuteAsync()`) instead of two (`ExecuteAsync()` and `ExecuteAsync(ITaskExecutionContext context)`).
+
+**Pros of Option B:**
+- ExecutionContext is explicitly visible in the method signature
```

Comment:
Just thought of one other factor: historically we have seen some trouble with folks stashing objects passed to a task for lifetimes longer than a single task execution--the `Log` object has particularly been problematic here. Passing it as a method argument may help discourage that (though obviously it doesn't make it impossible).